src/libraries/System.Text.Json/src/System/Text/Json/Nodes/JsonArray.IList.cs

```diff
@@ -114,6 +116,116 @@ public void RemoveAt(int index)
             DetachParent(item);
         }
 
+        /// <summary>
+        ///   Removes all the elements that match the conditions defined by the specified predicate.
+        /// </summary>
+        /// <param name="match">The predicate that defines the conditions of the elements to remove.</param>
+        /// <returns>The number of elements removed from the <see cref="JsonArray"/>.</returns>
+        /// <exception cref="ArgumentNullException">
+        ///   <paramref name="match"/> is <see langword="null"/>.
+        /// </exception>
+        public int RemoveAll(Func<JsonNode?, bool> match)
+        {
+            if (match == null)
+            {
+                ThrowHelper.ThrowArgumentNullException(nameof(match));
+            }
+
+            if (_list is not { } list)
+            {
+                return 0;
+            }
+
+            // Algorithm adapted from List<T>.RemoveAll()
```

Comment:
Rather than adapting the algorithm, is it easier/simpler just to use the underlying RemovalAll? For example:

```
list.RemoveAll(node => {
    if (match(node))
    {
        DetachParent(node);
        return true;
    }
    else
    {
        return false;
    }
});
```

I admit it's a bit strange to have a side effect on the matching function, but it makes this much much shorter (and also avoids a maintenance problem if List's implementation ever changes for some reason -- we get the benefit for free)

Reply:
This will create a lambda object capturing `match` for every `RemoveAll` call. Is the performance overhead anticipated? Usually I try to minimize object allocations.

Reply:
_Personally_ I wouldn't be worried about the single allocation here -- since this API tends to be more allocation heavy I don't imagine it's as critical. I'll let the API owners here decide what they want to do since ultimately they're the ones who have to maintain this code!

Reply:
> this API tends to be more allocation heavy

I agree with you that the API is to be decided by owners but I don't quite get that it is "more alloaction heavy". The `RemoveAll` methods from `List<T>` and my impl here are allocation-free, aren't they?

Reply:
Agree that incurring a single allocation isn't the end of the world in the context of JsonNode APIs specifically. By delegating to the `List.RemoveAll` method there are theoretical performance benefits in manipulating the underlying array directly.

Reply:
> I don't quite get that it is "more alloaction heavy".

Indeed that was ambiguous -- my thought was the JsonNode APIs are generally expensive in that you're already allocating several objects for each underlying node in the JSON object. If allocations are a huge concern, you might be using something else in the first place.